bigframes/core/nodes.py

```diff
@@ -0,0 +1,233 @@
+# Copyright 2023 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from __future__ import annotations
+
+from dataclasses import dataclass, field
+import typing
+from typing import Optional, Tuple
+
+import pandas
+
+import bigframes.core.guid
+from bigframes.core.ordering import OrderingColumnReference
+import bigframes.core.window_spec as window
+import bigframes.dtypes
+import bigframes.operations as ops
+import bigframes.operations.aggregations as agg_ops
+
+if typing.TYPE_CHECKING:
+    import ibis.expr.types as ibis_types
+
+    import bigframes.core.ordering as orderings
+    import bigframes.session
+
+
+@dataclass(frozen=True)
+class BigFrameNode:
+    pass
```

Comment:
"pass" isn't necessary since there is a body to this class. Let's add a docstring (maybe here's where we explain the need to be hashable for all subclasses?)

```suggestion
```

Reply:
done